[ADD] inventory: format zebra labels (#9699)
odoo/documentation: content/applications/inventory_and_mrp/inventory/shipping_receiving/setup_configuration/zebra.rst

=========================
Zebra label configuration
=========================

.. |ZPL| replace:: :abbr:`ZPL (Zebra Programming Language)`

> In Odoo, labels printed in the Zebra Programming Language (ZPL) file format is designed to fit a

Reviewer: ```suggestion
In Odoo, labels printed in the Zebra Programming Language (ZPL) file format are designed to fit a
```

> four-by-six inch label. To resize or format text to fit a variety of |ZPL| label sizes,

Reviewer: ```suggestion
four-by-six inch label. To resize (or reformat) text to fit a variety of |ZPL| label sizes,
```

> :ref:`navigate to the ZPL label view <inventory/shipping_receiving/zpl-view>` and alter the |ZPL|

Reviewer: ```suggestion
:ref:`navigate to the ZPL label view <inventory/shipping_receiving/zpl-view>`, and alter the |ZPL|
```
SUGGESTION: added comma for enhanced sentence flow

code.

> Refer to the following sections for explanations and example code for frequently requested Zebra

Reviewer: ```suggestion
Refer to the following sections for explanations, and example code, for frequently requested Zebra
```
SUGGESTION: added commas for enhanced sentence flow

label customizations.

- :ref:`adjust margins <inventory/shipping_receiving/margin>`
- :ref:`enlarge/minimize barcodes <inventory/shipping_receiving/resize>`
- :ref:`rotate elements <inventory/shipping_receiving/rotate>`

.. _inventory/shipping_receiving/zpl-view:

Navigate to ZPL label view
==========================

To begin customizing a Zebra label in Odoo, turn on :ref:`developer mode <developer-mode>`, and on

> the Odoo's main dashboard, type in `Reports`. From the search results, choose

Reviewer: ```suggestion
the main Odoo dashboard, type `Reports`. From the search results that appear in the resulting pop-up window, choose
```

> :guilabel:`Settings/Technical/Reporting/Reports` to open the :guilabel:`Reports page`.

Reviewer: ```suggestion
:guilabel:`Settings/Technical/Reporting/Reports` to open the :guilabel:`Reports` page.
```

.. note::
   To manually navigate to the :guilabel:`Reports` page, go to :menuselection:`Settings app -->
   Technical --> Reporting: Reports`.

.. image:: zebra/search.png
   :align: center
   :alt: Show global search result for "Reports".

> On the :guilabel:`Reports` page, in the :guilabel:`Search...` bar, type `ZPL` to get a list of

Reviewer: ```suggestion
On the :guilabel:`Reports` page, in the :guilabel:`Search...` bar, type `ZPL`, and hit :kbd:`Enter`. Upon doing so, Odoo presents a list of
```

> available Zebra labels in Odoo. Select the desired Zebra label to modify.

Reviewer: ```suggestion
available Zebra labels in Odoo. Select the desired Zebra label from the list to modify it on a separate page.
```

.. figure:: zebra/qweb-views.png
   :align: center
   :alt: Show Qweb smart button on the Lot and Serial Number (ZPL) report.

   **Lot and Serial Number (ZPL)** report, highlighting the Qweb smart button.

.. note::
   Printable ZPL labels in Odoo:

   - finished product
   - :ref:`lot/serial number <inventory/shipping_receiving/lot-sn-labels>`
   - operation type
   - package barcode
   - :ref:`product label <inventory/shipping_receiving/product-labels>`
   - product packaging

> Next, select the :icon:`fa-code` :guilabel:`Qweb Views` smart button, and choose the desired label

Reviewer: ```suggestion
Next, click the :icon:`fa-code` :guilabel:`Qweb Views` smart button, and choose the desired label
```
Just a personal nitpick here -- I just feel that when a "button" is involved, the term "click" sounds better. When "options" are involved, the term "select" seems best. But that's just me...

> :doc:`view <../../../../../developer/reference/user_interface/view_records>`. On the view page, go

Reviewer: ```suggestion
:doc:`view <../../../../../developer/reference/user_interface/view_records>`. On the resulting view form, go
```

to the :guilabel:`Architecture` tab to view the |ZPL| code.

.. important::

>    To ensure the customization are not overwritten during an update, click the :icon:`fa-bug`

Reviewer: ```suggestion
   To ensure the customization is **not** overwritten during an update, click the :icon:`fa-bug`
```

>    :guilabel:`(bug)` icon on the view page. Then, select :guilabel:`View Metadata` option to open

Reviewer: ```suggestion
   :guilabel:`(bug)` icon on the view page. Then, select the :guilabel:`View Metadata` option from the resulting drop-down menu, in order to open
```

>    the pop-up window. Then, ensure the :guilabel:`No Update` field is set to :guilabel:`true

Reviewer: ```suggestion
   the :guilabel:`View Metadata` pop-up window. Then, ensure the :guilabel:`No Update` field is set to :guilabel:`true
```

>    (change)`.

Reviewer: ```suggestion
   (change)`. Click :guilabel:`Ok` to remove the :guilabel:`View Metadata` pop-up window.
```

.. image:: zebra/architecture.png
   :align: center
   :alt: Architecture tab in the view.

.. _inventory/shipping_receiving/margin:

Adjust margin
=============

Text gets cut off from standard |ZPL| labels printed in Odoo when the line exceeds fifty-five

> characters. To fit long product names or lot numbers on a single line, adjust the margin.

Reviewer: ```suggestion
characters. To fit long product names, or lot numbers, on a single line, adjust the margin.
```
SUGGESTION: added comma to enhance sentence flow

> To begin, navigate to the :ref:`ZPL code of the label <inventory/shipping_receiving/zpl-view>`. In

Reviewer: This is all happening in the 'Architecture' tab of the view form, correct? Might be worth considering a quick reminder about that, just in case the reader gets a bit lost. Totally optional, though. 👍

> the |ZPL| code, look for the `^FT` command, which specifies where to start placing the text or

Reviewer: Unfortunately, I'm having a hard time finding a `^FT` command...maybe I'm just missing it? 🤔 
![Screenshot 2024-06-07 092137](https://github.com/odoo/documentation/assets/73958186/20b7664d-c7a3-47ff-85cd-40bfe46eb82d)
Seeing a lot of '^FS` though

So, this might be worth a quick double-check during your revisions

Author: THANK YOU for pointing this out! I didn't think to check the lot number label! Added an important block to say that if they're customizing lot labels, they have to change the `^FO` value instead

> the |ZPL| code, look for the `^FT` command, which specifies where to start placing the text or
> graphic element on the label. The two numbers immediately following `^FT` define the x-coordinate

Reviewer: ```suggestion
the |ZPL| code, look for the `^FT` command, which specifies where to start placing the text, or
graphic element, on the label. The two numbers immediately following `^FT` define the x-coordinate
```
SUGGESTION: added commas for enhanced sentence pacing/flow

and y-coordinate in dots (:dfn:`similar to pixels for printers`) from the left and top margins.

.. example::

    The following is an example where the |ZPL| name gets cut off with Odoo's default |ZPL|

>     formatting. In the **Fixed** tab, the x-coordinate of the starting position of the label is

Reviewer: curious...would Fixed need a GUI here? Or do you think **bold** is the right move?

Author: Since Fixed is not part of the UI, I'll leave it bolded (:

    changed from `^FT100,80` to `^FT0,80`, to fit the entire name.

.. tabs::

    .. tab:: Default

        .. image:: zebra/default-margin.png
           :align: center
           :alt: Example barcode label with the product name cut off.

        **Code**:

        .. code-block:: xml

            ^XA^CI28
            ^FT100,80^A0N,40,30^FD[E-COM11] Cabinet with Doors (wood: Cherry, handles: brass)^FS
            ...
            ^XZ

    .. tab:: Fixed

        .. image:: zebra/fixed-margin.png
          :align: center
          :alt: Example barcode label with the product name margin adjusted to the left.

        **Code**:

        .. code-block:: xml

            ^XA^CI28
            ^FT0,80^A0N,40,30^FD[E-COM11] Cabinet with Doors (wood: Cherry, handles: brass)^FS
            ...
            ^XZ

.. _inventory/shipping_receiving/resize:

Resize barcode
==============

To adjust the size of the barcode to scale, begin by navigating to the :ref:`ZPL code of the label
<inventory/shipping_receiving/zpl-view>`. Look for the `^FO` command (typically in the third line),
which is the starting point of the margin for the barcode.

> The `^BY` command configures barcode size and takes three numbers: bar width, width of wide bars

Reviewer: ```suggestion
The `^BY` command configures barcode size, and takes three numbers: bar width, width of wide bars
```
SUGGESTION: added comma for enhanced sentence pacing

=========================
Zebra label configuration
=========================

.. |ZPL| replace:: :abbr:`ZPL (Zebra Programming Language)`

In Odoo, labels printed in the Zebra Programming Language (ZPL) file format are designed to fit a
four-by-six inch label. To resize (or reformat) text to fit a variety of |ZPL| label sizes,
:ref:`navigate to the ZPL label view <inventory/shipping_receiving/zpl-view>`, and alter the |ZPL|
code.

.. warning::
   When customizing code in Odoo, please note that upgrading the database to newer versions may
   break custom |ZPL| code. **Customers are responsible for maintaining their custom code**.

Refer to the following sections for explanations, and example code, for frequently requested Zebra
label customizations.

- :ref:`adjust margins <inventory/shipping_receiving/margin>`
- :ref:`enlarge/minimize barcodes <inventory/shipping_receiving/resize>`

> - :ref:`rotate elements <inventory/shipping_receiving/rotate>`

Reviewer 2: Just double-checking, should these be capitalized?

> - :ref:`adjust margins <inventory/shipping_receiving/margin>`
> - :ref:`enlarge/minimize barcodes <inventory/shipping_receiving/resize>`
> - :ref:`rotate elements <inventory/shipping_receiving/rotate>`

Author: ```suggestion
- :ref:`Adjust margins <inventory/shipping_receiving/margin>`
- :ref:`Enlarge/minimize barcodes <inventory/shipping_receiving/resize>`
- :ref:`Rotate elements <inventory/shipping_receiving/rotate>`
```

.. _inventory/shipping_receiving/zpl-view:

Navigate to ZPL label view
==========================

To begin customizing a Zebra label in Odoo, turn on :ref:`developer mode <developer-mode>`, and on
the main Odoo dashboard, type `Reports`. From the search results that appear in the resulting pop-up

> window, choose :guilabel:`Settings/Technical/Reporting/Reports` to open the :guilabel:`Reports`

Reviewer 3: I _believe_ there are spaces between the slash marks in the search box navigational address
```suggestion
window, choose :guilabel:`Settings / Technical / Reporting / Reports` to open the :guilabel:`Reports`
```

> window, choose :guilabel:`Settings / Technical / Reporting / Reports` to open the :guilabel:`Reports`
> page.

Reviewer 2: Small adjustment so this line isnt over 100 chars
```suggestion
window, choose :guilabel:`Settings / Technical / Reporting / Reports` to open the
:guilabel:`Reports` page.
```

.. note::
   To manually navigate to the :guilabel:`Reports` page, go to :menuselection:`Settings app -->
   Technical --> Reporting: Reports`.

.. image:: zebra/search.png
   :align: center
   :alt: Show global search result for "Reports".

> On the :guilabel:`Reports` page, in the :guilabel:`Search...` bar, type `ZPL`, and hit "Enter". Upon
> doing so, Odoo presents a list of available Zebra labels in Odoo. Select the desired Zebra label
> from the list to modify it on a separate page.

Reviewer 2: Suggestion to use the :kbd: role for pressing keys :) 

Side note, need to look into styling for this!
```suggestion
On the :guilabel:`Reports` page, in the :guilabel:`Search...` bar, type `ZPL`, and hit :kbd:`Enter`.
Upon doing so, Odoo presents a list of available Zebra labels in Odoo. Select the desired Zebra
label from the list to modify it on a separate page.
```

.. note::
   Printable ZPL labels in Odoo:

   - :ref:`lot/serial number <inventory/shipping_receiving/lot-sn-labels>`
   - operation type
   - package barcode
   - :ref:`product label <inventory/shipping_receiving/product-labels>`
   - product packaging

>    - finished product (to see this, **must** install Odoo *Manufacturing*)

Reviewer 3: missing subject, this might be simpler
```suggestion
   - finished product (Odoo *Manufacturing* app required)
```

Next, click the :icon:`fa-code` :guilabel:`Qweb Views` smart button, and choose the desired label
:doc:`view <../../../../../developer/reference/user_interface/view_records>`.

.. figure:: zebra/qweb-views.png
   :align: center
   :alt: Show Qweb smart button on the Lot and Serial Number (ZPL) report.

   **Lot and Serial Number (ZPL)** report, highlighting the Qweb smart button.

On the resulting view form, go to the :guilabel:`Architecture` tab to view the |ZPL| code.

.. important::

> 

Reviewer 2: Remove empty line
```suggestion
```

.. image:: zebra/architecture.png
   :align: center
   :alt: Architecture tab in the view.

.. _inventory/shipping_receiving/margin:

Adjust margin
=============

Text gets cut off from standard |ZPL| labels printed in Odoo when the line exceeds fifty-five
characters. To fit long product names, or lot numbers, on a single line, adjust the margin.

To begin, navigate to the :ref:`ZPL code of the label <inventory/shipping_receiving/zpl-view>` in
the :guilabel:`Architecture` tab. In the |ZPL| code for product labels, look for the `^FT` command,
which specifies where to start placing the text, or graphic element, on the label. The two numbers
immediately following `^FT` define the x-coordinate and y-coordinate in dots (:dfn:`similar to
pixels for printers`) from the left and top margins.

.. important::
   When customizing lot/serial number labels, look for the `^FO` command, instead of `^FT`.

.. example::

    The following is an example where the product's name gets cut off with Odoo's default |ZPL|
    formatting. In the **Fixed** tab, the x-coordinate of the starting position of the label is
    changed from `^FT100,80` to `^FT0,80`, to fit the entire name.

> .. tabs::
> 
>     .. tab:: Default
> 
>         .. image:: zebra/default-margin.png
>            :align: center
>            :alt: Example barcode label with the product name cut off.
> 
>         **Code**:
> 
>         .. code-block:: xml
> 
>             ^XA^CI28
>             ^FT100,80^A0N,40,30^FD[E-COM11] Cabinet with Doors (wood: Cherry, handles: brass)^FS
>             ...
>             ^XZ
> 
>     .. tab:: Fixed
> 
>         .. image:: zebra/fixed-margin.png
>           :align: center
>           :alt: Example barcode label with the product name margin adjusted to the left.
> 
>         **Code**:
> 
>         .. code-block:: xml
> 
>             ^XA^CI28
>             ^FT0,80^A0N,40,30^FD[E-COM11] Cabinet with Doors (wood: Cherry, handles: brass)^FS
>             ...
>             ^XZ

Reviewer 2: Alignment and suggestion to use the term "Modified" instead of "Fixed" since [tab headings are not translated](https://www.odoo.com/documentation/17.0/contributing/documentation/rst_cheat_sheet.html#content-tabs)
```suggestion
.. tabs::

   .. tab:: Default

      .. image:: zebra/default-margin.png
         :align: center
         :alt: Example barcode label with the product name cut off.

      **Code**:

      .. code-block:: xml

         ^XA^CI28
         ^FT100,80^A0N,40,30^FD[E-COM11] Cabinet with Doors (wood: Cherry, handles: brass)^FS
         ...
         ^XZ

   .. tab:: Modified

      .. image:: zebra/fixed-margin.png
         :align: center
         :alt: Example barcode label with the product name margin adjusted to the left.

      **Code**:

      .. code-block:: xml

         ^XA^CI28
         ^FT0,80^A0N,40,30^FD[E-COM11] Cabinet with Doors (wood: Cherry, handles: brass)^FS
         ...
         ^XZ
```

.. _inventory/shipping_receiving/resize:

Resize barcode
==============

To adjust the size of the barcode to scale, begin by navigating to the :ref:`ZPL code of the label
<inventory/shipping_receiving/zpl-view>` in the :guilabel:`Architecture` tab. Look for the `^FO`
command (typically in the third line), which is the starting point of the margin for the barcode.

The `^BY` command configures barcode size, and takes three numbers: bar width, width of wide bars
relative to narrow bars, and bar height. By default, |ZPL| code in Odoo uses `^BY3`, setting the bar
width to three dots, a typical size that is easy for barcode scanners to read.

.. example::
   To shrink the barcode to scale, `^BY3` is reduced to `^BY2`.

> .. tabs::
> 
>     .. tab:: Default
> 
>         .. image:: zebra/normal-barcode.png
>            :align: center
>            :alt: Example barcode label.
> 
>         **Code**:
> 
>         .. code-block:: xml
> 
>             ^XA^CI28
>             ...
>             ^FO100,160^BY3
>             ...
>             ^XZ
> 
>     .. tab:: Shrunken
> 
>         .. image:: zebra/shrink-barcode.png
>           :align: center
>           :alt: Example barcode label with the barcode size reduced.
> 
>         **Code**:
> 
>         .. code-block:: xml
> 
>             ^XA^CI28
>             ...
>             ^FO100,160^BY2
>             ...
>             ^XZ

Reviewer 2: Alignment and heading :)
```suggestion
.. tabs::

   .. tab:: Default

      .. image:: zebra/normal-barcode.png
         :align: center
         :alt: Example barcode label.

      **Code**:

      .. code-block:: xml

         ^XA^CI28
         ...
         ^FO100,160^BY3
         ...
         ^XZ

   .. tab:: Modified

      .. image:: zebra/shrink-barcode.png
         :align: center
         :alt: Example barcode label with the barcode size reduced.

      **Code**:

      .. code-block:: xml

         ^XA^CI28
         ...
         ^FO100,160^BY2
         ...
         ^XZ
```

.. _inventory/shipping_receiving/rotate:

Rotate elements
===============

To rotate elements in |ZPL|, begin by navigating to the :ref:`ZPL code of the label
<inventory/shipping_receiving/zpl-view>` in the :guilabel:`Architecture` tab.

The `^BC` command's first parameter (:dfn:`information that affects the behavior of the command`)
defines the rotation of an item, which can be:

- `N`: display normally
- `R`: rotate 90 degrees
- `I`: rotate 180 degrees
- `B`: rotate 270 degrees

.. example::
   To rotate the barcode, `^BCN` is changed to `^BCB`.

> .. tabs::
> 
>     .. tab:: Default
> 
>         .. image:: zebra/lot.png
>            :align: center
>            :alt: Example barcode label.
> 
>         **Code**:
> 
>         .. code-block:: xml
> 
>             ^XA^CI28
>             ...
>             ^BCN,100,Y,N,N
>             ...
>             ^XZ
> 
>     .. tab:: Rotated
> 
>         .. image:: zebra/rotate.png
>           :align: center
>           :alt: Example barcode label with the barcode rotated.
> 
>         **Code**:
> 
>         .. code-block:: xml
> 
>             ^XA^CI28
>             ...
>             ^BCB,100,Y,N,N
>             ...
>             ^XZ

Reviewer 2: ```suggestion
.. tabs::

   .. tab:: Default

      .. image:: zebra/lot.png
         :align: center
         :alt: Example barcode label.

      **Code**:

      .. code-block:: xml

         ^XA^CI28
         ...
         ^BCN,100,Y,N,N
         ...
         ^XZ

   .. tab:: Modified

      .. image:: zebra/rotate.png
         :align: center
         :alt: Example barcode label with the barcode rotated.

      **Code**:

      .. code-block:: xml

          ^XA^CI28
          ...
          ^BCB,100,Y,N,N
          ...
          ^XZ
```
